### ISSUE TITLE
Need way to automatically show usage if required params are missing

### ISSUE THREAD

**Issue by lukehutch** [cbeust/jcommander#337]
Currently if you specify `help = true` in a parameter, exceptions are suppressed. However, you then have to manually check for required parameters not being set.

It would be great if there were a way to prevent exceptions from being thrown on missing required parameters, but print the usage text instead. Maybe in addition to adding `help = true` to a help parameter, you could also specify `usage = true`, and if a parameter is missing, even if the help option was not given on the commandline, JCommander could set the help parameter value to true, so that the usage text is displayed?

**Comment by lukehutch** [comment]
The reason the current pattern is problematic is that you don't have a `JCommander` reference if the constructor throws an exception, so you can't call the `usage()` method if there is a missing parameter, unless there is a `help = true` parameter. But if there is a `help = true` parameter, you don't get built-in required parameter checking. It's not possible currently to get both built-in required parameter checking *and* be able to display usage info if there is a missing required parameter.

**Comment by cbeust** [comment]
How about adding the JCommander instance to the ParameterException?


-- 
Cédric

**Comment by lukehutch** [comment]
@cbeust  That wouldn't work -- it is the constructor that throws the exception. The instance is never constructed.

**Comment by lukehutch** [comment]
@cbeust However, you could add a `usage()` method to the exception. That would solve the problem.

**Comment by xpromache** [comment]
Hello,

I vote for having the ParameterException contain the JCommander instance. That would also help me with sub-commands, it allows knowing where the exception has been thrown  from. 

As for the constructor throwing the exception - it's because you do the parsing in the constructor. I would deprecate that usage. Construct and then parse is cleaner. 
I had a quick look at the standard Java XML parsers - they do the same - first construct the parser and then parse.

**Comment by lukehutch** [comment]
Correct. I have run into this exact sort of issue a couple of times before in my own code -- having constructors throw exceptions can lead to lots of nasty situations like this.

Adding methods to exceptions is indeed a bit weird (nobody ever thinks to go looking for methods or fields in exceptions, so the API is not obvious without consulting the documentation), so changing the behavior to construct then parse does make more sense.

**Comment by cbeust** [comment]
Agreed, doing the parse in the constructor was a poor decision, I'll deprecate it.

**Comment by lukehutch** [comment]
Yeah, I remember concluding years ago that doing a lot of work in a constructor was almost always a bad idea. I still do this sort of thing a lot, but I often have to re-learn this lesson the hard way :-)

**Comment by xpromache** [comment]
For my case, I'm using sub-commands and I need to have the JCommander instance that generated the exception in order to print the appropriate message.

For example when you run "git commit --invalidoption" you get a usage message for the "git commit" not for git. 
And while you are at it, can you please  also make Jcommander.addCommand(...) return the child JCommand instance created so I can add subcommands to it? Currently I have to use getCommands to get the Map and then I can get to the child JCommand.

**Comment by cbeust** [comment]
I just deployed 1.67 with this change.

`ParameterException` now has both `getJCommander()` and `usage()`.

Note that this only work if you parse with `parse()` instead of the constructor.

**Comment by lukehutch** [comment]
Thanks. The latest version in Maven Central is 1.64, and the latest release on GitHub is 1.66. Also, neither the GitHub page nor the `jcommander.org` site has documentation on how this works (although I figured it out from the source -- in particular, it wasn't obvious to me that you needed to call `JCommander(Object argObject)`, then `parse(String[] args)` ). What I have is:

```java
        JCTest jcTest = new JCTest();
        JCommander jc = new JCommander(jcTest);
        try {
            jc.parse(args);
        } catch (ParameterException e) {
            e.usage();
            System.exit(1);
        }
```

**Comment by cbeust** [comment]
Correct.

Note also that the constructor is deprecated, you should  use the builder instead (I just updated the documentation to reflect that):

```
    JCTest jcTest = new JCTest();
    JCommander jc = JCommander.newBuilder().addObject(jcTest).build();
```

**Comment by lukehutch** [comment]
@cbeust when you say "I just deployed 1.67 with this change", where did you deploy it to? That version is not tagged on github, and is not released in Maven Central. Just wondering if you're using something other than github or Maven Central as your main release channel.

**Comment by cbeust** [comment]
JCenter: http://jcenter.bintray.com/com/beust/jcommander/1.67/

**Comment by lukehutch** [comment]
With 1.67 I get:

`JCommander.newBuilder()` -> `The type JCommander.Builder is not visible`

**Comment by cbeust** [comment]
Oops, my mistake. Fixed and deployed in 1.68.
